### ISSUE TITLE
API Create Candidate Recorders in Undebate

### ISSUE THREAD

**Issue by ddfridley** [EnCiv/undebate-ssp#127]
similar to #106  
use `import { undebatesFromTemplateAndRows } from 'undebate'` to create the viewer and recorders.  See 
[app/socket-apis/create-moderator-recorder.js](https://github.com/EnCiv/undebate-ssp/blob/main/app/socket-apis/create-moderator-recorder.js) for how example use.

Pull the moderators recording from the database, to use as the moderator for the candidate recorders, and for the viewer.
Create the rowObjs by pulling data from electionObj.candidates

Create test tests for this

Create a web-component and use npm run dev to manually test the the recorders and viewer.  See [app/web-components/qa-create-moderator-recorder.js](https://github.com/EnCiv/undebate-ssp/blob/main/app/web-components/qa-create-moderator-recorder.js) for an example. And the iota.json entry with the path: qa-create-moderator-recorder

**Comment by ddfridley** [comment]
@ice1080 
I am making a change to the electionObj schema.  The moderator.submissions object will have the docs of the moderator's recordings.  You need the speaking array.  Here is an example doc that can be found in app/socket-apis/tests/get-election-docs.  I will update issue #46 when I get the chance.  The `Object` text below is something that jest creates, you can ignore it.

For each candidate, we need to create viewer, recorder, and submission objects.  This can be done by adding functions to the mergeOps object at the bottom of app/socket-apis/get-election-doc.  It seems like mergeOps should be factored out into something more general - but we need to focus on getting something working for now. 

```
Object {
              "_id": "628c73daf2014b3f4c5da4ee",
              "description": "Election document #4",
              "subject": "Election document",
              "userId": "628d0a2afacbb605f4d8e6ac",
              "webComponent": Object {
                "moderator": Object {
                  "recorders": Object {
                    "628d076dcf19df5aa438c07a": Object {
                      "_id": "628d076dcf19df5aa438c07a",
                      "bp_info": Object {
                        "office": "Moderator",
                      },
                      "component": Object {
                        "component": "undebateCreator",
                      },
                      "description": "Moderator Recorder for #4",
                      "parentId": "628c73daf2014b3f4c5da4ee",
                      "subject": "Moderator Recorder for #4",
                      "userId": "628d0a2afacbb605f4d8e6ac",
                      "path": "/some-moderator-specific-recorder-path",
                    },
                  },
                  "submissions": Object {
                    "628d2d25c945f836b8be0901": Object {
                      "_id": "628d2d25c945f836b8be0901",
                      "component": Object {
                        "component": "MergeParticipants",
                        "participant": Object {
                          "listening": "https://res.cloudinary.com/hf6mryjpf/video/upload/v1566510649/5d5b73c01e3b194174cd9b92-0-seat2.webm",
                          "name": "david",
                          "speaking": Array [
                            "https://res.cloudinary.com/hf6mryjpf/video/upload/v1566510654/5d5b73c01e3b194174cd9b92-1-speaking.webm",
                            "https://res.cloudinary.com/hf6mryjpf/video/upload/v1566510659/5d5b73c01e3b194174cd9b92-2-speaking.webm",
                            "https://res.cloudinary.com/hf6mryjpf/video/upload/v1566510665/5d5b73c01e3b194174cd9b92-3-speaking.webm",
                          ],
                        },
                      },
                      "description": "Moderator Recording for #4",
                      "parentId": "628d0b225f7a7746488c0bff",
                      "subject": "Moderator Recording for #4",
                      "userId": "628d0a2afacbb605f4d8e6ac",
                    },
                  },
                  "viewers": Object {
                    "628d0b225f7a7746488c0bff": Object {
                      "_id": "628d0b225f7a7746488c0bff",
                      "bp_info": Object {
                        "office": "Moderator",
                      },
                      "description": "Moderator Viewer for #4",
                      "parentId": "628c73daf2014b3f4c5da4ee",
                      "subject": "Moderator Viewer for #4",
                      "userId": "628d0a2afacbb605f4d8e6ac",
                      "webComponent": Object {
                        "webComponent": "CandidateConversation",
                      },
                      "path": "/some-moderator-specific-viewer-path",
                    },
                  },
                },
                "webComponent": "ElectionDoc",
              },
            }
```

**Comment by ice1080** [comment]
One more question, I noticed that in your example above, I believe I would have to know the object id of the submissions object to get to the speaking array, does that sound correct?
```
electionObj.moderator.submissions['<submissions-id-here?>'].component.participant.speaking
```
unless I am getting the electionObj and the electionDocs confused?

**Comment by ddfridley** [comment]
> @ddfridley I noticed that create-moderator-recorder.js creates the `msgs` list that it gets from `reasonsNotReadyForModeratorRecorder`, but then never does anything with that or sends it to the callback or anything. Is this intentional? Should I be doing the same or is there some behavior we would like to add here?

The APIs just return undefined if there is an error. My thinking was that the function would get exported and used in the UI somewhere, someday.

**Comment by ddfridley** [comment]
> One more question, I noticed that in your example above, I believe I would have to know the object id of the submissions object to get to the speaking array, does that sound correct?
> 
> ```
> electionObj.moderator.submissions['<submissions-id-here?>'].component.participant.speaking
> ```
> 
> unless I am getting the electionObj and the electionDocs confused?

Yes. 
In app/lib/get-election-status-methods is
```
const getLatestIota = iotas => {
    if (!iotas) return undefined
    if (typeof iotas !== 'object') return undefined
    const latest = Object.values(iotas).reduce((latest, obj) => (idCompare(latest._id, obj._id) < 0 ? obj : latest), {
        _id: '',
    })
    if (!latest._id) return undefined
    return latest
}
```
getLatestIota(electionObj.moderator.submissions) would get the object you should use.  I would say use that function, but you are working on the api side rather than the browser side.  Id say export it out of there, and then import that function into what you are doing.

**Comment by ice1080** [comment]
I see within `sendModeratorInvite` that we do this:
```
import getElectionStatusMethods from '../lib/get-election-status-methods'
...
const electionMethods = getElectionStatusMethods(undefined, electionObj)
...
electionMethods.getLatestIota(...)
```
So I believe it's already exported, and I'll use the same thing!

**Comment by ice1080** [comment]
It looks like we'll need a specific intro video for candidates. The intro video for the moderator one specifically mentions the moderator. It might be possible to create a video that would work for both and replace the moderator one too, though I'm not familiar enough with the whole process just yet to say for sure.

**Comment by ddfridley** [comment]
For now use the existing one, and lets come up with a rough script for a new one.  Maybe we'll find a volunteer to redo both.

**Comment by ddfridley** [comment]
> I see within `sendModeratorInvite` that we do this:
> 
> ```
> import getElectionStatusMethods from '../lib/get-election-status-methods'
> ...
> const electionMethods = getElectionStatusMethods(undefined, electionObj)
> ...
> electionMethods.getLatestIota(...)
> ```
> 
> So I believe it's already exported, and I'll use the same thing!

That will work, but it's extra overhead to put into a server side API function, if all that's needed is that one function.  But, go for what's easiest to get working for now.

**Comment by icdt-ihudson** [comment]
Sounds good, I will make a stretch goal to put that function in the server side somewhere.

**Comment by ice1080** [comment]
Sounds good, I will make a stretch goal to put that function in the server side somewhere. (sorry if double commented, used the wrong account)

**Comment by ice1080** [comment]
I'm not sure I fully understand what is needed relative to how the create moderator recorder does things. I see that it uses the `moderatorViewerRecorder`, which requires specific data and functions to work with the undebate function `undebatesFromTemplateAndRows`. But will we be doing the same thing with a `candidateViewerRecorder` and also passing it to `undebatesFromTemplateAndRows` or will we be modifying some other data?

**Comment by ice1080** [comment]
Another question, possibly related, I'm having trouble finding where in all the json is the list of candidates that we'll be updating. I see it in iotas.json, but perhaps I'm just having a hard time wrapping my head around the relation between the data in a file like moderatorViewerRecorder and the data in iotas.json, if that makes sense.

**Comment by ddfridley** [comment]
#46 is the best example of candidates
You could also go to the candidate table in storybook and look at the election object that gets generated. Next week we'll be able to generate data in the database. 
We can talk about the other on Tuesday.

**Comment by ice1080** [comment]
Ya I definitely understand where it lives in the electionObj, but is this API just modifying the electionObj directly or is it calling something else like how create moderator recorder calls the function from undebate? That's fine if you'd like to wait till Tuesday for now.

**Comment by ddfridley** [comment]
Here's something that may help. I have been adding stuff to the new [schema wiki page](https://github.com/EnCiv/undebate-ssp/wiki/electionObj-data-schema).  If there is anything I should explain more - please let me know.

the create-candidate-recorders api, will use getElectionDocById.call(this,id,cb) to get the electionDoc.  The function returns the whole doc, including the outer Iota structure. Where as an electionObj would be the webComponent of the elecctionDoc. And the terminology is a little fuzzy here, what gets returned will have moderator recorders, viewers, and invitations merged in.  They are not part of the electionDoc, they are their own mongo docs, but the mergeOps functions at the bottom of get-elections-docs.js will pull in the related docs.  

Part of what needs to be done for this issue is to create more mergeOp functions for adding the invitations, submissions, and recorders under each candidate.   And, I'm saying merge, but in postgres I think this is like a join. 

So the createCandidateRecorders API won't make changes to the electionObj/doc itself. I will just call undebatesFromTemplateAndRows that will cause new recorder and viewer docs to be created. 

Does this help?

**Comment by ice1080** [comment]
Sounds good. Right now both create-moderator-recorder and create-candidate-recorder were using this to get the electionObj, so I'll update it to use getElectionDocById instead.
```
const iota = await Iota.findOne({ _id: Iota.ObjectId(id) })
if (!iota) {
  if (cb) cb()
  return
}
const electionObj = iota.webComponent
```
The rest of what you explained makes sense. I'll look at how the moderator has been done in the mergeOps, that also explains why I wasn't clear on where the candidates were, as mergeOps had not yet been updated. I believe most of my questions have now been answered. I'll dig into that and see if that answers the rest of them. Thanks!

**Comment by ddfridley** [comment]
Ya. To create the candidate recorders the moderator submissions are needed but won't be there before the mergeOps.

**Comment by ice1080** [comment]
Ok I believe I am getting somewhat close on this one. A few more questions regarding get-election-docs for what I'm hung up on currently:
1. For each of the candidate child objects (recorder, viewer, submissions), will the parentId of each iota be the candidate id or the elctionDoc id?
If this value is the candidateId, then we need to have a candidate iota in the db, which I'm not exactly sure what this will look like (I don't believe I've seen this yet in the code or in any of my mongo digging). Additionally, this wouldn't match how the moderator's objects of the same types store their data (might not matter).
If instead this value is the electionDoc id then we need some other way to determine which candidate this is for.
I guessing you've already figured this out in undebate, I just can't seem to find it correctly in the db.
1. Within the mergeOps, I am using `child?.bp_info?.office !== 'Moderator'`, but is that the correct way to determine that an object is for a candidate instead of a moderator?

I believe those are the only questions holding me up, the rest of my questions are lower priority and I'll ask them on Tuesday.
Thanks!

**Comment by ddfridley** [comment]
1) parentId of the recorder iota doc will be the viewer doc.
ParentId of the viewer iota doc will be the the election doc. This is how it is for the moderator so you can see how the merge Ops works for those.  bp_info for the recorder should include the unique-id of the candidate, and the office form the candidate table. You have to add the info to bp_info when creating the undebate. 

2) I think there is more detail to it that you will figure out from 1. Each candidate in election obj needs to get their own  submission - based on the unique_id

As for the viewers, all candidates for the same office will get the same viewer. I think we will need to add an offices property to the electionObj. The key will be the name of the office. The object will have a viewers property.  I'll add something to the schema tomorrow.

**Comment by ice1080** [comment]
I keep feeling like I have a grasp on things in this conversation and then I try to implement and my understanding seems to disappear ha. Maybe just typing this out will help me figure out what I'm missing, so I'll do that separately and potentially post it in here if I feel like others could benefit from it. As a note, when you said above that the "parentId of the recorder iota doc will be the viewer doc", I believe you meant "of the recording" instead? I'm seeing that in the tests of get-election-docs and the parentId of the recorder is like the viewer - the election doc. Does that sound correct?

**Comment by ice1080** [comment]
And I could be mistaken but I believe the mergeOps of the candidates will be a good bit different than the moderator as the obj path for the moderator is as simple as "webComponent.moderator.submissions" for example, but for the candidates I have to add each item to each of the candidates in "webComponent.candidates.<every candidate id>.submissions"

**Comment by ice1080** [comment]
I just pushed the candidatesRecorder method in mergeOps, I need to test it for multiple candidates but I believe it should work. That one was the easy one since the bp_info contains the candidate id in unique_id (as mentioned). We can talk tomorrow about my questions regarding the viewer/submissions merge ops

**Comment by ice1080** [comment]
@ddfridley I thought that I just about had the candidate submissions, but I got stuck on one thing that might just be a unit test thing. Here is my code in mergeOps:
```
    candidatesSubmissions(dst, root, child, iotas, usedIndexes) {
        if (!(child?.component?.component === 'MergeParticipants')) {
            return false
        }
        const candidateName = root?.webComponent?.component?.participant?.name
        if (!candidateName) return false
        const candidate = root?.webComponent?.candidates.find(cand => cand.name === candidateName)
        if (!candidate) return false
        intoObjOfDocsAtObjPathMergeDoc(dst, `webComponent.candidate.${candidate.uniqueId}.submissions`, child)
        return true
    },
```
The problem I see (at least in the unit tests) is that the candidate.name has not yet been assigned in the webComponent. I know that root.webComponent is the electionObj, and I know that eventually the electionObj has the filled out candidates list, I'm curious when/how that happens? 
It might just be that the get-election-docs unit test does not have enough data in the setup iotas to fill it out. Would that data live in the top level electionDoc?

**Comment by ddfridley** [comment]
The candidate names will get into the electionObj when someone uploads the CSV file.  You might have to manufacture some of your own records for the unit test. the \_\_test\_\_/find-and-set-election-doc has  mockElection with candidate data.

If you want to see it in action you can `npm run dev` with the latest build, and browse to localhost:3011/undebates and join and create an undebate.  And then look at the database. (have you installed mongo compass? )

I think you basically have it  - but here's a clarification.
```
candidateName=child?.component?.participant?.name 

candidate=Object.values(root?.webComponent?.candidates || {}).find(cand=>cand.name===candidateName)
```
child is the iota (of iotas) that the mergeOp is trying to merge into dst.

**Comment by ddfridley** [comment]
FYI the bp_info will be in the recording. That record is created in [undebate app/components/lib/create-participant](
https://github.com/EnCiv/undebate/blob/d5539bcef7ea43f093435708d60760768ca57868/app/components/lib/create-participant.js#L72)

**Comment by ice1080** [comment]
I believe I've found a problem with get-election-docs for what we're trying to do. We last talked about doing something like this for the candidate recordings (MergeParticipants):
```
ElectionDoc {
  ...
  office: {
    ...
    viewer: {
      ...
      submissions: {
        <candidate-recordings>
      }
    }
  }
}
```
But it seems like that won't work because of this line in `intoDstOfRootMergeChildrenOfParentFromIotasMarkingUsedIndexs`: 
```
if (parentId !== child.parentId) continue
```
Does that sound correct? I will have more time to dig into this on either Sunday or Monday to learn more and try to find a way to fix this

**Comment by ddfridley** [comment]
I __think__ I had this problem with the moderator recorder. That's why moderatorViewers in mergeOps returns the function to call, with the viewer as the parentId, but still the same dst and root.  Does that help?

**Comment by ice1080** [comment]
I'll dig in more to confirm. Maybe I had something wrong in my test, but my merge op was using the same method.

**Comment by ice1080** [comment]
All right, I had somehow modified what I had copied in the function call. I had `getId(child.parentId)` for the parent Id because I hadn't initially understood that function, but now I have it as `getId(child._id)` just like moderatorViewer. Cleaning up a few things then will have you do a quick code review to see if what I've got for the election docs is what you're looking for.

**Comment by ice1080** [comment]
Alright @ddfridley, if you happen to have time before tomorrow's meeting, can you look at these two files on my fork? Let me know if there's anything blatantly missing from that? I'm still working on create-candidate-recorder.js so it's not quite ready yet, but the below should be ready, assuming that's what we want our office viewer to look like.
https://github.com/ice1080/undebate-ssp/blob/api-create-candidate-recorders%23127/app/socket-apis/get-election-docs.js
https://github.com/ice1080/undebate-ssp/blob/api-create-candidate-recorders%23127/app/socket-apis/__tests__/get-election-docs.js
One question I have is if the viewer object should be an object of id objects, or just a straight object? As it was merged with my code, you'll notice that the destination of the candidate submissions has to be this:
```
`webComponent.offices.${candidate.office}.viewer.${child.parentId}.submissions`,
```
Once we confirm that's how we want it I'll get the other test updated as well. I just now noticed that jest will update the inline snapshots like it updates files so that should save me some time from manually updating like I did this last one ha.

**Comment by ddfridley** [comment]
In https://github.com/ice1080/undebate-ssp/blob/api-create-candidate-recorders%23127/app/socket-apis/get-election-docs.js
I don't think line 266 is necessary.  I don't think office is used. 

The rest we need to discuss.  I was thinking that the submissions should go under the candidates uniqueId. The UI needs to know if the candidate has submitted anything. `webComponent.candidates.[uniqueId].submissions`

**Comment by ice1080** [comment]
Agreed on line 266, that was one I missed from an old test version locally.
Ok we can talk about the submissions tomorrow. Last time we talked I thought you had said the submissions were going to go under the office viewer, but maybe I just misunderstood and you only meant that for the parentId of the submissions.

**Comment by ddfridley** [comment]
Here is an example of the moderator submisission, (recording, participan record, ...)
```
{
  "_id": { "$oid": "62b2812bd404ff5e34d4d217" },
  "component": {
    "component": "UndebateCreator",
    "participants": {
      "moderator": {
        "speaking": [
          "https://res.cloudinary.com/huu1x9edp/video/upload/q_auto/v1614893716/5d5b74751e3b194174cd9b94-1-speaking20210304T213504684Z.mp4",
          "https://res.cloudinary.com/huu1x9edp/video/upload/q_auto/v1614893717/5d5b74751e3b194174cd9b94-3-speaking20210304T213516602Z.mp4",
          "https://res.cloudinary.com/huu1x9edp/video/upload/q_auto/v1614893718/5d5b74751e3b194174cd9b94-5-speaking20210304T213517487Z.mp4",
          "https://res.cloudinary.com/huu1x9edp/video/upload/q_auto/v1614893719/5d5b74751e3b194174cd9b94-6-speaking20210304T213517927Z.mp4",
          "https://res.cloudinary.com/huu1x9edp/video/upload/q_auto/v1614894042/5d5b74751e3b194174cd9b94-1-speaking20210304T214041075Z.mp4",
          "https://res.cloudinary.com/huu1x9edp/video/upload/q_auto/v1614894043/5d5b74751e3b194174cd9b94-2-speaking20210304T214041861Z.mp4"
        ],
        "name": "David Fridley, EnCiv",
        "listening": "https://res.cloudinary.com/huu1x9edp/video/upload/q_auto/v1614893720/5d5b74751e3b194174cd9b94-0-listening20210304T213518628Z.mp4",
        "agenda": [
          ["How this works", "Placeholder"],
          ["What is your favorite color?"],
          ["Do you have a pet?"],
          ["Should we try to fix income inequality?"],
          ["Make your closing - to the audience"],
          ["Thank you"]
        ],
        "timeLimits": [15, 60, 60, 60, 60]
      },
      "human": {
        "listening": { "round": 0, "seat": "speaking" },
        "name": "Bill Smith"
      }
    }
  },
  "webComponent": {
    "webComponent": "Undebate",
    "logo": "undebate",
    "instructionLink": "",
    "participants": {},
    "opening": { "noPreamble": false },
    "closing": {
      "thanks": "Thank You.",
      "iframe": {
        "src": "https://docs.google.com/forms/d/e/1FAIpQLSchcQjvnbpwEcOl9ysmZ4-KwDyK7RynwJvxPqRTWhdq8SN02g/viewform?embedded=true",
        "width": 640,
        "height": 1550
      }
    }
  },
  "bp_info": {
    "office": "Moderator",
    "election_date": "03/21/2021",
    "candidate_name": "Bill Smith",
    "last_name": "Smith",
    "unique_id": "62b2812bd404ff5e34d4d216",
    "candidate_emails": ["billsmith@gmail.com"],
    "party": "",
    "election_source": "CodeForAmerica.NAC"
  },
  "subject": "Moderator-Candidate Recorder",
  "description": "A Candidate Recorder for the undebate: Moderator",
  "path": "/country:us/organization:cfa/office:moderator/2021-03-21-recorder-62b2812bd404ff5e34d4d216",
  "parentId": "621e8aefe7db9b6338d0ab74"
}

```

**Comment by ice1080** [comment]
> The candidate names will get into the electionObj when someone uploads the CSV file. You might have to manufacture some of your own records for the unit test. the __test__/find-and-set-election-doc has mockElection with candidate data.
> 
> If you want to see it in action you can `npm run dev` with the latest build, and browse to localhost:3011/undebates and join and create an undebate. And then look at the database. (have you installed mongo compass? )

@ddfridley I tried creating my own locally and got this when I click the submit button. It appears that it's getting created when I refresh the page, but just wanted to note it somewhere.
```
 Failed to load resource: the server responded with a status of 404 (Not Found)
site.webmanifest:1 Manifest: Line: 1, column: 1, Syntax error.
2bconsole.js:18 2022Jul05 23:02:09 browser upsert: no updates until state has been populated
subscribe-election-info.js:15 merge called with Objectid: "62c51751ca00756b8062e299"webComponent: "ElectionDoc"[[Prototype]]: Object
subscribe-election-info.js:20 upsert called with ObjectelectionName: "Isaac Undebate"[[Prototype]]: Object
subscribe-election-info.js:20 upsert called with ObjectorganizationName: "ice1080"[[Prototype]]: Object
election-component.js:50 Uncaught TypeError: onDone is not a function
    at onDone (election-component.js:50:35)
    at handleOnClick (submit.js:15:13)
    at HTMLUnknownElement.callCallback (react-dom.development.js:188:1)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:237:1)
    at invokeGuardedCallback (react-dom.development.js:292:1)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:306:1)
    at executeDispatch (react-dom.development.js:389:1)
    at executeDispatchesInOrder (react-dom.development.js:414:1)
    at executeDispatchesAndRelease (react-dom.development.js:3278:1)
    at executeDispatchesAndReleaseTopLevel (react-dom.development.js:3287:1)
react-dom.development.js:327 Uncaught TypeError: onDone is not a function
    at onDone (election-component.js:50:35)
    at handleOnClick (submit.js:15:13)
    at HTMLUnknownElement.callCallback (react-dom.development.js:188:1)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:237:1)
    at invokeGuardedCallback (react-dom.development.js:292:1)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:306:1)
    at executeDispatch (react-dom.development.js:389:1)
    at executeDispatchesInOrder (react-dom.development.js:414:1)
    at executeDispatchesAndRelease (react-dom.development.js:3278:1)
    at executeDispatchesAndReleaseTopLevel (react-dom.development.js:3287:1)
site.webmanifest:1          Failed to load resource: the server responded with a status of 404 (Not Found)
```

**Comment by ice1080** [comment]
Regarding create-candidate-recorder, I'm not sure I'll be able to get much further without understanding what goes into undebatesFromTemplateAndRows. It's definitely not a subset of the electionObj, but I've tried doing something similar to the moderator-viewer-recorder and overwriting what I know and that doesn't seem to successfully create the CandidateConversation either. It's possible that I don't have the right test data setup yet, but when I try to create a new undebate from scratch with the dev server I can't seem to get far enough along with all the errors and such. 
Just confirming, within the candidateViewer the webComponent.participants.moderator is actually the moderator and won't be the candidate right? And same for candidateRecorder, the component.participants.moderator is not the candidate right?

Here's the basics of what I've got right now, and the candidateViewerRecorder is basically the same as the moderatorViewerRecorder.
```
            const moderatorComponent = getLatestIota(electionObj.moderator.submissions).component
            const speaking = moderatorComponent.participant.speaking
            speaking.unshift(introVideo)

            candidateViewerRecorder.candidateRecorder.component.participants.moderator.speaking = speaking
            candidateViewerRecorder.candidateRecorder.component.participants.moderator.listening =
                moderatorComponent.participant.listening
            candidateViewerRecorder.candidateRecorder.component.participants.human.name = candidate.name

            candidateViewerRecorder.candidateViewer.webComponent.participants.moderator.speaking = speaking
            candidateViewerRecorder.candidateViewer.webComponent.participants.moderator.name =
                electionObj.moderator.name
            candidateViewerRecorder.candidateViewer.bp_info.name = candidate.name
            candidateViewerRecorder.candidateViewer.bp_info.email = candidate.email
            candidateViewerRecorder.candidateViewer.bp_info.uniqueId = candidate.uniqueId

            const inRowObjs = [{ Seat: candidate.office, Name: candidate.name, Email: candidate.email }]
            const { rowObjs, messages } = await undebatesFromTemplateAndRows(candidateViewerRecorder, inRowObjs)
```

**Comment by ddfridley** [comment]
Its SO GOOD to have you asking about this.  It's forcing me to write stuff down and make it clearer. 

undebatesFromTemplateAndRows is going to use the template (the candidateViewerRecorder), and the rowObjs (an array of row objects), to create one recorder Iota for each candidate, and one viewer Iota for each office.   (The moderator example was a degenerate case with 1 and 1. )

undebatesFromTemplateAndRows can be called to create the data initially, update data, or add new candidates and offices. It also makes sure that it is updating an existing viewer Iota if one already exists, and updating candidate recorder Iotas if they already exist. 

Inside the template (the viewer recorder template) the main function that will need to be customized for different situations is: `getIotaPropertyFromCSVColumn`.   Each function in this object, will make a piece of data that is used in the viewer and or recorder Iota.  The function gets called with each rowObj as a parameter, so it may get data out of the rowObj, or it may hard code the response, or do something else.

What you have looks good for setting up the moderator for the candidates.

But setting bp_info for the candidateViewer is done by undebatesFromTemplateAndRows so that shouldn't be done here.  Also, I see a terminology problem - the candidateViewer terms is misleading - it would be more descriptive as candidate__s__ viewer, or office viewer, because there is one viewer for all the candidates in an office.  But lets not try to fix that here.

To be more specific, inRowObjs is more like
```
const inRowObjs=Object.values(electionObj.candidates).map(candidate=>{ Seat: candidate.office, Name: candidate.name, Email: candidate.email, unique_id: candidate.uniqueId })
```
uniqueId is needed, it wasn't needed for the moderator but it is for he candidates.


This line needs a .slice() on it to make a copy of the speaking array, before the introVideo is added.  Otherwise, the introVideo would be in the moderatorComponent as well.  I started out as an embedded C programmer and can tell a longer story on all this that makes javascript object and array instantiation clearer. 
```
const speaking = moderatorComponent.participant.speaking.slice()
```

If in app/socket-apis/subscribe-election-info.js
you change
```
            sendInvitation() {
                dispatch(
                     ....
            },
```
to 
```
    sendInvitation() {
        window.socket.emit('create-moderator-recorder', id)
    }
```
Then it will create the moderator viewer and recorder.  We can worry about sending it later. And I'm thinking maybe we should have 2 buttons, one to create, and another one to send and in between a way for the election administrator to review the moderator recorder.  But that should be a separate issue.

**Comment by ice1080** [comment]
> The function gets called with each rowObj as a parameter, so it may get data out of the rowObj, or it may hard code the response, or do something else.

For this section, here's what I've got for that, and the questions I have:
```
const getIotaPropertyFromCSVColumn = {
    type_name: () => 'organization',
    type_id: row => 'cfa',
    election_date: row => '03/21/2021',
    candidate_name: row => row['Name'],
    office: row => row['Seat'],
    candidate_emails: row => [row['Email']],
    unique_id: row => row['unique_id'],
    set: {
        unique_id: (row, val) => (row['unique_id'] = val),
    },
    party: row => '',
    viewer_url_name: () => 'viewer_url',
    recorder_url_name: () => 'recorder_url',
    url_update_name: name => name + '_updated',
    lastname: row => row['Name'].split(' ').reduce((acc, word) => word, ''), // the last word in full name will be the last name
    election_source: () => 'CodeForAmerica.NAC',
}
```
These so far are basically the same as moderator. I believe that the main important one we'll need to update is the election_date and then next would be type_name, type_id, and election_source, though those might not be as important. From what I'm seeing in the template file(s), it appears that we'd have to pass in more to the inRowObjs to be able to figure out things like the election_date; is that correct?

-------------

> This line needs a .slice() on it to make a copy of the speaking array

Ya that makes total sense and I should have noticed that it'd be pointing to the same object in memory without the slice(). My work is mainly in groovy which oftentimes handles makes copies for you so takes me a minute to get used to javascript again.

-------------

Regarding the change for `sendInvitation`, I need to understand that a bit more. Are you saying that you want me to commit the change from dispatch to emit? Or that is just for my local testing?

**Comment by ice1080** [comment]
Btw, I haven't tested the links yet, but with what we've got so far, I'm at least getting this from my qa component now:
```
{
  "rowObjs": [
    {
      "Seat": "President of the U.S.",
      "Name": "Sarah Jones",
      "Email": "sarahjones@mail.com",
      "unique_id": "61e76bbefeaa4a25840d85d0",
      "recorder_url": "https://cc.enciv.org/country:us/organization:cfa/office:president-of-the-u-s/2021-03-21-recorder-61e76bbefeaa4a25840d85d0",
      "recorder_url_updated": "yes",
      "viewer_url": "https://cc.enciv.org/country:us/organization:cfa/office:president-of-the-u-s/2021-03-21",
      "viewer_url_updated": "yes"
    },
    {
      "Seat": "President of the U.S.",
      "Name": "Michael Jefferson",
      "Email": "mikejeff@mail.com",
      "unique_id": "61e76bfc8a82733d08f0cf12",
      "recorder_url": "https://cc.enciv.org/country:us/organization:cfa/office:president-of-the-u-s/2021-03-21-recorder-61e76bfc8a82733d08f0cf12",
      "recorder_url_updated": "yes",
      "viewer_url": "https://cc.enciv.org/country:us/organization:cfa/office:president-of-the-u-s/2021-03-21",
      "viewer_url_updated": "yes"
    },
    {
      "Seat": "President of the U.S.",
      "Name": "Isaac Hudson",
      "Email": "foo@bar.com",
      "unique_id": "62c7ae332cdb0433eefbb60a",
      "recorder_url": "https://cc.enciv.org/country:us/organization:cfa/office:president-of-the-u-s/2021-03-21-recorder-62c7ae332cdb0433eefbb60a",
      "recorder_url_updated": "yes",
      "viewer_url": "https://cc.enciv.org/country:us/organization:cfa/office:president-of-the-u-s/2021-03-21",
      "viewer_url_updated": "yes"
    }
  ],
  "messages": [
    "created recorder President of the U.S., /country:us/organization:cfa/office:president-of-the-u-s/2021-03-21, /country:us/organization:cfa/office:president-of-the-u-s/2021-03-21-recorder-61e76bbefeaa4a25840d85d0",
    "created recorder President of the U.S., /country:us/organization:cfa/office:president-of-the-u-s/2021-03-21, /country:us/organization:cfa/office:president-of-the-u-s/2021-03-21-recorder-61e76bfc8a82733d08f0cf12",
    "created recorder President of the U.S., /country:us/organization:cfa/office:president-of-the-u-s/2021-03-21, /country:us/organization:cfa/office:president-of-the-u-s/2021-03-21-recorder-62c7ae332cdb0433eefbb60a"
  ]
}
```

**Comment by ddfridley** [comment]
> These so far are basically the same as moderator. I believe that the main important one we'll need to update is the election_date and then next would be type_name, type_id, and election_source, though those might not be as important. From what I'm seeing in the template file(s), it appears that we'd have to pass in more to the inRowObjs to be able to figure out things like the election_date; is that correct?

Yes.  It would be better to refactor it so we don't have to pass in things like election_source in every row, but not for this moment.  type_name is going to be a challenge.  It needs to be unique for each organization, but we don't have any logic for preventing election admins from naming their organizations uniquely. Thoughts?  One is to use the userId. It's ugly and longer than it needs to be.  For now, just use the first letter of each word in the organizationName or something easy but not necessarily perfect. 

Regarding send invitation - you can commit it.  

> I'm at least getting this from my qa component now
Looks good.

**Comment by ice1080** [comment]
The next issue I'm seeing is that the new viewer objects seem to not have parentIds attached to them. I don't see us doing this in moderator-viewer-recorder, but I do see us unit testing that there exists a viewer iota that has the correct parentId. I'm guessing this then gets set somewhere in undebate? I can't seem to figure out where in undebate that would be occurring at and what I'm missing from mine to make it occur.

**Comment by ddfridley** [comment]
In app/socket-api/create-moderator-recorder.js life 92 is
```
moderatorViewerRecorder.candidateViewer.parentId = id
```

**Comment by ice1080** [comment]
:facepalm: I must be too tired, I should have noticed that. So sorry! Going to bed after this lol.

I tested the recorder link for my user and see it successfully gathering the recording. It froze on upload but that could be because I didn't have the parentId on the viewer when I was testing this. Now that I've got that, I also see my unit tests getting further so should be able to finish those up soon. I think this one's _finally, actually_ almost there ha.

**Comment by ddfridley** [comment]
Do you have cloudinary setup?  uploads go to cloudinary.  I usually create a heroku app, and add cloudinary to it (free), and then get the CLOUDINARY_URL out of the env from heroku and put it in .bashrc

**Comment by ddfridley** [comment]
speaking of .slice()
In create-moderator-recorder.js right after
```
import { Iota } from 'civil-server'
import { undebatesFromTemplateAndRows } from 'undebate'

import moderatorViewerRecorder from '../components/lib/moderator-viewer-recorder'
```
I should have done
```
import {cloneDeep} from "lodash"
```
and then inside createModeratorRecorder should have done
```
const newModeratorViewerRecorder=cloneDeep(moderatorViewerRecorder)
```
and then mutate the newModeratorViewerRecorder everywhere else.

Thanks for helping me realize this.  Not sure if clonedeep works on functions though. Make it a class?  I'll think about it tomorrow.

**Comment by ice1080** [comment]
Question on the usage of this whole thing, I'm guessing the election admin will be performing an action that will cause this to get called? Specifically I'm asking, the individual candidates will not be performing the action that calls this, correct?

**Comment by ddfridley** [comment]
Correct. The admin does it about once or maybe 2 or 3 times per election.

**Comment by ice1080** [comment]
Perfect, had to update some things around that. Another question, will the candidates have a "message" property, like the moderator has?

**Comment by ice1080** [comment]
One downside to using a real database with shared data between tests is that it can easily cause test pollution. When I run the create-candidate-recorder test on its own, everything passes. When I run the create-moderator-recorder test on its own, everything passes. When I run all the unit tests, they oftentimes cause test pollution. To help with some of this, I changed the testDoc id and in the afterAll sections I added 
```
    await Iota.deleteMany({})
```
which ensures that when the tests are run serially they all pass. To run them serially, use this:
```
npm run test -- --runInBand
```
Then they run serially which takes a bit longer but allows them to all pass. If they are not run serially, then the parallel features of jest causes one of the tests to finish, deleting all the iotas, while the other one is still checking iotas.

If you want to force all test runs to be serial because of this, you can always update all the test scripts in package.json. You can also update the documentation to state this, up to you how you'd like to handle it.

A true solution to this would be to create a mongo transaction per test file and rollback the transaction at the end of each test so that there's no pollution, but I don't know if MongoModels has that option.

I'm finishing up a few more tests and then should have a PR ready for this tonight for review and any changes you'd like.

**Comment by ddfridley** [comment]
No message for candidates.

**Comment by ddfridley** [comment]
I have had the problem with multiple db tests in parallel. My solution was to give each iota its own unique _id and make sure that my queries wouldn't catch things outside the scope of my test.

You can use 
'node node_modules/civil-server/dist/tools/mongo-id.js' from the command line to generate unique -ids.

**Comment by ddfridley** [comment]
It was also an interesting challenge with testing socket-apis but eventually I figured out how to open a new server for each socket test running in parallel.

**Comment by ice1080** [comment]
Unfortunately the pollution is not caused by me inserting objects with the same ids; I think the pollution is actually caused by the moderator test and candidates test having the same viewer. I'm not sure exactly how this is happening though. But ya it seems to be failing here, because the viewerId is somehow the same, so it's getting the moderator Undebate object too. :thinking: 
{{{
    const iotas = await Iota.find({ 'webComponent.webComponent': 'Undebate', parentId: viewerId })
}}}

**Comment by ice1080** [comment]
One thing I just noticed while testing this, it seems like when I'm recording my videos that the other candidate (you in this case) gets one extra recording to say "Vote for me". Does that sound accurate? I'm also not seeing a script so maybe I'm still missing something. Anyway, I'll create the PR and let you run it locally to see if there's some things I'm missing.

**Comment by ddfridley** [comment]
I gave you a bad example to start with.
I did this in the tests for create-moderator-recorder and create-candidate-recorders
```
// a clone with a unique _id so multiple jest tests can run in parallel
const testDoc = cloneDeep(iotas.filter(iota => iota.subject === 'Undebate SSP Test Iota')[0])
testDoc._id = ObjectID('62c8f80565128d16c8a62590')
```
Now they run in parallel.


> when I'm recording my videos that the other candidate (you in this case) gets one extra recording to say "Vote for me". Does that sound accurate?
Maybe. The moderator always speaks first to introduce the undebate, and last to thank all the voters for participating".  Are you thinking "I" am a candidate, or am "I" the moderator?

**Comment by ice1080** [comment]
Ok I'll update the tests for that, good to know. That makes sense yet again that I would have to clone it. I again forgot that those would be pointing at the same reference, so when I updated one of the ids, it was updating both. 

Maybe my moderator videos from the db are old links and pointing at you acting as a candidate because the videos are definitely answering questions and not asking them. I'll try again to start a new undebate from scratch and see if I can get further. I'll probably work on that tomorrow, as I'm going on the undebates list now.

**Comment by ddfridley** [comment]
FYI - I pushed the clone changes I made to the branch.

⁣Get BlueMail for Android ​
